src/SignalR/clients/ts/signalr/src/HubConnection.ts

```diff
@@ -769,6 +769,7 @@ export class HubConnection {
 
                 if (this._connectionState !== HubConnectionState.Reconnecting) {
                     this._logger.log(LogLevel.Debug, "Connection left the reconnecting state during reconnect attempt. Done reconnecting.");
+                    this._completeClose();
```

Comment:
We do the same thing in the .NET client
https://github.com/dotnet/aspnetcore/blob/a3316184b39c0f4e0f4e5646ff4a9f2e369f6d0b/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L1398

Reply:
That's not exactly the same thing. The analogous code in the JS client is the following:

https://github.com/dotnet/aspnetcore/blob/brecon/tsdisconnect/src/SignalR/clients/ts/signalr/src/HubConnection.ts#L269-L280 


I think the change to have `_startInternal()`  call `_stopPromiseResolver()` in its catch block before returning is enough to fix this.

Reply:
> That's not exactly the same thing.

I think it is, I linked to the Reconnect code in HubConnection.cs which aligns with the Reconnect code being changed in HubConnection.ts, you're linking to stop which is different.

> I think the change to have `_startInternal()` call `_stopPromiseResolver()` in its catch block before returning is enough to fix this.

It isn't, they are at two different layers, the `_stopPromiseResolver` code is for closing the HttpConnection. HubConnection will still be in a bad state unless we call `_completeClose`.

Reply:
There's no exactly equivalent code to the following in the TS client.

https://github.com/dotnet/aspnetcore/blob/a3316184b39c0f4e0f4e5646ff4a9f2e369f6d0b/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L1398

In .NET, we have to handle the OperationCanceledException after cancelling StopCts before we can complete the Task returned by Stop(). In TS, `clearTimeout(this._reconnectDelayHandle)` means no more code runs in the reconnect loop, not even error handling code, so we can return a completed promise from stop() directly which we do.

> It isn't, they are at two different layers, the `_stopPromiseResolver` code is for closing the HttpConnection. HubConnection will still be in a bad state unless we call `_completeClose`.

`this._completeClose()` is called before `return Promise.resolve()` when there's a `this._reconnectDelayHandle`. I'm talking about the same layer. There's a `_startInternal()` [in HttpConnection too](https://github.com/dotnet/aspnetcore/blob/614908248d20c4a1572d1a3bbc7a997d12be9d01/src/SignalR/clients/ts/signalr/src/HttpConnection.ts#L294-L299). A failed start of `HttpConnection` shouldn't leave anything in a bad state.

Reply:
Talked offline. This is important not for when the reconnect delay is canceled but for when a reconnect attempt fails simultaneously with a call to HubConnection.stop().